### PR TITLE
expression: implement vectorized evaluation for `builtinFormatSig`

### PR DESCRIPTION
**Pull request by mmyj**
PCP #12106

### What problem does this PR solve? 
implement vectorized evaluation for `builtinFormatSig` , for #12106

### What is changed and how it works?
```
goos: linux
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinStringFunc/builtinFormatSig-VecBuiltinFunc-3              2834            441774 ns/op           96545 B/op       5120 allocs/op
BenchmarkVectorizedBuiltinStringFunc/builtinFormatSig-NonVecBuiltinFunc-3           2229            545643 ns/op           96544 B/op       5120 allocs/op
BenchmarkVectorizedBuiltinStringFunc/builtinFormatSig-VecBuiltinFunc#01-3           2737            449215 ns/op          124608 B/op       5120 allocs/op
BenchmarkVectorizedBuiltinStringFunc/builtinFormatSig-NonVecBuiltinFunc#01-3        2139            553661 ns/op          124609 B/op       5120 allocs/op
PASS
ok      github.com/pingcap/tidb/expression      5.596s

```

### Check List 

Tests 

- Unit test

### CODE REVIEW

File: expression/builtin_string_vec.go [hunk 1]
```diff
@@ -1646,11 +1646,110 @@ func (b *builtinBinSig) vecEvalString(input *chunk.Chunk, result *chunk.Column)
...
+	n := input.NumRows()
+	buf0, err := b.bufAllocator.get(types.ETReal, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf0)
+	if err := b.args[0].VecEvalReal(b.ctx, input, buf0); err != nil {
+		return err
+	}
+	buf1, err := b.bufAllocator.get(types.ETInt, n)
```
Comment (XuHuaiyu) [inline expression/builtin_string_vec.go]: There are so many duplicate codes. Can we use a variable like `xStrBuf []string` to store the intermediate result for the two different types?

File: expression/builtin_string_vec_test.go [hunk 2]
```diff
@@ -218,6 +218,10 @@ var vecBuiltinStringCases = map[string][]vecExprBenchCase{
 	ast.InsertFunc: {
 		{retEvalType: types.ETString, childrenTypes: []types.EvalType{types.ETString, types.ETInt, types.ETInt, types.ETString}, geners: []dataGenerator{&randLenStrGener{10, 20}, &rangeInt64Gener{-10, 20}, &rangeInt64Gener{0, 100}, &randLenStrGener{0, 10}}},
 	},
+	ast.Format: {
+		{retEvalType: types.ETString, childrenTypes: []types.EvalType{types.ETDecimal, types.ETInt}, geners: []dataGenerator{&rangeDecimalGener{}, &rangeInt64Gener{2, 5}}},
```
Comment (XuHuaiyu) [inline expression/builtin_string_vec_test.go]: We can use `rangeInt64Gener{0, 30}` to cover more cases.

File: expression/builtin_string_vec.go (hunk 1 above)
Comment (mmyj) [reply expression/builtin_string_vec.go]: I will try it.

File: expression/builtin_string_vec_test.go (hunk 2 above)
Comment (mmyj) [reply expression/builtin_string_vec_test.go]: I will fix it.

File: expression/builtin_string_vec.go [hunk 3]
```diff
@@ -1770,11 +1770,71 @@ func (b *builtinBinSig) vecEvalString(input *chunk.Chunk, result *chunk.Column)
...
+		}
+		buf1.MergeNulls(buf0)
+		f64s := buf0.Float64s()
+		for i := 0; i < n; i++ {
+			if buf1.IsNull(i) {
+				continue
+			}
+			xStrBuf[i] = fmt.Sprintf(strconv.FormatFloat(f64s[i], 'f', -1, 64))
+		}
+	}
```
Comment (qw4990) [inline expression/builtin_string_vec.go]: How about creating a new function for the logic like `evalNumDecArgsForFormat` then another format function `builtinFormatWithLocaleSig` can also use it.

Comment (mmyj) [reply expression/builtin_string_vec.go]: 👌

Comment (mmyj) [reply expression/builtin_string_vec.go]: I will try it tomorrow.

**Comment by qw4990** [comment]
@mmyj Friendly ping, please resolve conflicts at your convenience.

**Comment by mmyj** [comment]
> @mmyj �Friendly ping, please resolve conflicts at your convenience.

I will resolve it tonight.

**Pull request merged**